### PR TITLE
Record count caching optimization

### PR DESCRIPTION
**Pull request by ssuvorov-fls**
fixes https://github.com/OHDSI/WebAPI/issues/1308

### CODE REVIEW

**Comment by ssuvorov-fls** [comment]
@anthonysena Could you review changes?

**Comment by alondhe** [comment]
@ssuvorov-fls -- I'd like to make this process optional, as not everyone uses Atlas. Can you add a parameter to the achilles function call, perhaps named something like "optimizeAtlasCache" and add the details to the meta-tagging?

**Comment by ssuvorov-fls** [comment]
@alondhe agree, parameter "optimizeAtlasCache" was added

**Comment by anthonysena** [comment]
@alondhe - just following up on this one to see if there are any other changes requested? This looks good from my quick review.

**Comment by alondhe** [comment]
Looks good to me now. @anthonysena -- want to take 1 last look?

**Comment by alondhe** [comment]
@ssuvorov-fls -- can you create a new function ("optimizeAtlasCache") that will allow this query to be run separately after an Achilles execution? So the achilles function can call it if the user sets the parameter to TRUE, but they can also run it later if they had set it to FALSE.

**Comment by ssuvorov-fls** [comment]
@alondhe Hello, could you review changes?

File: R/Achilles.R [hunk 1]
```diff
@@ -1006,6 +1029,73 @@ dropAllScratchTables <- function(connectionDetails,
...
+optimizeAtlasCache <- function(connectionDetails,
+                               resultsDatabaseSchema,
+                               vocabDatabaseSchema = resultsDatabaseSchema,
+                               outputFolder = "output",
+                               sqlOnly = FALSE,
+                               verboseMode = TRUE,
+                               tempAchillesPrefix = "tmpach") {
+
+  # Log execution --------------------------------------------------------------------------------------------------------------------
+
```
Comment (alondhe) [inline R/Achilles.R]: add logic to create the outputFolder if it doesn't exist

File: R/Achilles.R [hunk 2]
```diff
@@ -631,6 +642,18 @@ achilles <- function (connectionDetails,
...
+    optimizeAtlasCacheSql <- optimizeAtlasCache(connectionDetails = connectionDetails,
+                                                resultsDatabaseSchema = resultsDatabaseSchema,
+                                                vocabDatabaseSchema = vocabDatabaseSchema,
+                                                outputFolder = outputFolder,
+                                                sqlOnly = sqlOnly,
+                                                verboseMode = verboseMode,
+                                                tempAchillesPrefix = tempAchillesPrefix)
+
+    achillesSql <- c(achillesSql, optimizeAtlasCacheSql)
+}
```
Comment (alondhe) [inline R/Achilles.R]: Can you add a tab here to fix the indent?

File: R/Achilles.R [hunk 3]
```diff
@@ -631,6 +642,18 @@ achilles <- function (connectionDetails,
                                 achillesTables = unique(achillesTables))
   }
   achillesSql <- c(achillesSql, indicesSql)
+
```
Comment (alondhe) [inline R/Achilles.R]: Can you add a section comment like this:

`# Optimize Atlas Cache -----------------------------------------`

File: R/Achilles.R [hunk 4]
```diff
@@ -1006,6 +1029,73 @@ dropAllScratchTables <- function(connectionDetails,
...
 
+#' Optimize atlas cache
+#'
+#' @details
+#' Post-processing, optimize data for atlas cache in separate table to help performance.
+#'
+#' @param connectionDetails                An R object of type \code{connectionDetails} created using the function \code{createConnectionDetails} in the \code{DatabaseConnector} package.
+#' @param resultsDatabaseSchema		       Fully qualified name of database schema that we can write final results to. Default is cdmDatabaseSchema.
+#'                                         On SQL Server, this should specifiy both the database and the schema, so for example, on SQL Server, 'cdm_results.dbo'.
+#' @param vocabDatabaseSchema
```
Comment (alondhe) [inline R/Achilles.R]: Can you fulfill the comment for this parameter:

`#' @param vocabDatabaseSchema		           String name of database schema that contains OMOP Vocabulary. Default is cdmDatabaseSchema. On SQL Server, this should specifiy both the database and the schema, so for example 'results.dbo'.
`

File: inst/sql/sql_server/analyses/create_result_concept_table.sql [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+      SELECT
+        CAST(concept_id AS VARCHAR) ancestor_id,
+        CAST(concept_id AS VARCHAR) descendant_id
+      FROM @vocabDatabaseSchema.concept c
+  ), counts AS (
+  SELECT stratum_1 concept_id, MAX (count_value) agg_count_value
+  FROM @resultsDatabaseSchema.achilles_results
+  WHERE analysis_id IN (2, 4, 5, 201, 225, 301, 325, 401, 425, 501, 505, 525, 601, 625, 701, 725, 801, 825,
+  826, 827, 901, 1001, 1201, 1425, 1801, 1825, 1826, 1827, 2101, 2125, 2301)
+      /* analyses:
```
Comment (alondhe) [inline inst/sql/sql_server/analyses/create_result_concept_table.sql]: Can you remove these multi-line comments? They can interfere with some SqlRender patterns

File: R/Achilles.R (hunk 1 above)
Comment (ssuvorov-fls) [reply R/Achilles.R]: fixed

File: R/Achilles.R (hunk 2 above)
Comment (ssuvorov-fls) [reply R/Achilles.R]: fixed

File: R/Achilles.R (hunk 3 above)
Comment (ssuvorov-fls) [reply R/Achilles.R]: fixed

File: R/Achilles.R (hunk 4 above)
Comment (ssuvorov-fls) [reply R/Achilles.R]: fixed

File: inst/sql/sql_server/analyses/create_result_concept_table.sql (hunk 5 above)
Comment (ssuvorov-fls) [reply inst/sql/sql_server/analyses/create_result_concept_table.sql]: fixed

File: R/Achilles.R (hunk 1 above)
Comment (alondhe) [reply R/Achilles.R]: I'm not sure this works. When I tested on my machine, naming an outputFolder that didn't exist, the logging failed. I'm thinking instead using this at the top of the function:

```
  if (!dir.exists(outputFolder)) {
    dir.create(path = outputFolder, recursive = TRUE)
  }
```

File: inst/sql/sql_server/analyses/create_result_concept_table.sql (hunk 5 above)
Comment (alondhe) [reply inst/sql/sql_server/analyses/create_result_concept_table.sql]: Sorry -- meant can you remove the comments from the body of the SQL script? We can simply document this in the readme.

File: R/Achilles.R (hunk 1 above)
Comment (ssuvorov-fls) [reply R/Achilles.R]: I don't understand
I tried this on my machine and it worked
```optimizeAtlasCache(connectionDetails,
         resultsDatabaseSchema="results",
         vocabDatabaseSchema = "dbo",
         sqlOnly=FALSE,
         outputFolder="test_output")   ```

File: inst/sql/sql_server/analyses/create_result_concept_table.sql (hunk 5 above)
Comment (ssuvorov-fls) [reply inst/sql/sql_server/analyses/create_result_concept_table.sql]: done

**Comment by alondhe** [comment]
Thanks @ssuvorov-fls! Merging

**Pull request merged**